### PR TITLE
Tidy up GroupInfo.info

### PR DESCRIPTION
**Pull request by seanh**
Part of https://github.com/hypothesis/lms/issues/1839

`GroupInfo.info` is one of three JSON columns that we added to LMS before we realised that using JSON in Postgres is quite a pain due to lack of mutation tracking and lack of defaults and constraints.

The other two are `ApplicationInstance.settings` and `Course.settings`, which we've made safe (ish) and the model code nice enough by introducing a custom SQLAlchemy type `ApplicationSettings` with safe `get()` and `set()` methods and a docstring warning you not to bypass them: https://github.com/hypothesis/lms/pull/1909

`ApplicationSettings` doesn't apply to the `GroupInfo.info` column. `ApplicationSettings` is for maintaining a two-level dict in which each top-level item is a sub-dict and then the sub-dicts contain settings (and the settings themselves should be simple values like strings and booleans, not further dicts and lists, or you'll have to deal with mutation tracking issues). `GroupInfo.info` is different: it's used to store two keys: `"instructors"` is a list (defaults to `[]`) and `"type"` is a string (defaults to `"course_group"`).

I don't think it's worthwhile trying to devise a custom SQLAlchemy type for `GroupInfo.info`'s case at this time, so this PR just tidies a little bit of Python code mess that's been made while fixing unforeseen JSON-handling issues:

1. Rename `GroupInfo.info` (which is the raw dict which may be `None` or may be a dict with either of the `"instructors"` or `"type"` fields missing) to `GroupInfo._info` to discourage accessing it directly

2. Delete an unnecessary `__init__` method that was setting default values for `GroupInfo._info` and the `"instructors"` and `"type"` fields it contains. Setting defaults in `__init__` only works when creating new `GroupInfo`'s, it doesn't always work when reading them from the DB, and nothing prevents `info` from being `None` or being a dict with no `"instructors"` or `"type"` in the DB.

   The convenience `@property`'s `instructors` and `type` provide ways of accessing instructors and type with default values that work in both cases as long as you always go through those properties, so the `__init__` simply isn't needed.

3. Add a docstring warning that modifications won't be saved

`GroupInfo` now implements a fairly reasonable approach to "how to deal with JSON" in SQLAlchemy, as far as handling JSON's lack of constraints and defaults goes:

1. Name the JSON column itself with a leading `_` to warn people off it
2. Add a private `@property` for safely accessing the column (with defaulting to `{}`), like the `_safe_info` property below
3. Add getter and setter `@property`'s for each piece of public info that you're storing in your JSON column. These properties should be based on the private `_safe_foo` property, and then should provide their own defaults for the specific items that they get.

Note that this **does nothing to help with mutation tracking**. If you do `group_info.instructors.append(something)` that **will not be saved**. (Actually, sometimes it will be saved, and sometimes it won't, and a change elsewhere in the code can trip it from being saved to not being saved.) Instead you have to either do `group_info.instructors = group_info.instructors + [something]` or, if you do mutate it in place, you have to remember to call `group_info._info.changed()`. I've added a docstring warning about this.

It's disappointing how ugly this is and how much boilerplate Python code is needed. You've got the `_info` attribute itself and the `_safe_info` property, and then a getter and a setter property for each individual item in the dict. Even after adding `_info` and `_safe_info`, it's actually more additional lines of Python code and complexity per field than if each field had its own non-JSON column. I think this behaviour probably _can_ somehow be wrapped up in reusable code so that we don't need so much junk in every model class that has a JSON column. But that still wouldn't fix the mutation issue.

My honest opinion for the future: we should probably avoid JSON in future except when it's really needed, for something that actually wouldn't work or would be truly painful to do with non-JSON columns. For most cases, for example if we're just using JSON to try to avoid adding multiple Postgres columns / SQLAlchemy attributes to the model, JSON looks to be more trouble than it's worth, it will add considerably _more_ code noise to your model not less and it will make you deal mutation, defaults and constraints yourself instead of Postgres doing it. It also forces you to use Postgres's JSON querying syntax, which you probably aren't familiar with, rather than plain SQL if you're every trying to inspect the DB directly.

```python
class GroupInfo:

    _info = sa.Column("info", MutableDict.as_mutable(JSONB))

    @property
    def _safe_info(self):
        if self._info is None:
            self._info = {}

        return self._info

    @property
    def instructors(self):
        return self._safe_info.get("instructors", [])

    @instructors.setter
    def instructors(self, new_instructors):
        self._safe_info["instructors"] = new_instructors

    @property
    def type(self):
        return self._safe_info.get("type", None)

    @type.setter
    def type(self, new_type):
        self._safe_info["type"] = new_type
```

### CODE REVIEW

**Comment by jon-betts** [comment]
As for the more subjective stuff I have a few comments. This is here mostly for colour, please don't read into it too much.

In the case of `type` it's probably true that a DB column would be easier. In the case of `instructors` it's definitely not. This would be a table with relations and all that fun jazz, with multiple queries and all sorts of other aspects.

This definitely pulls more stuff from the DB and into the code, but that's kind of the feature here. This is / should be, the conscious motivation. We explicitly want to move certain aspects out of database control and into code control. Not because it makes for fewer lines of code or what-have-you, but because we want to gain freedoms from the DB. So to the observation "this has moved things out of the DB and created code", the answer is "yes"?

That doesn't mean we can't refine the art and cut the waffle down though.

The statement about Postgres JSON syntax being unfamiliar doesn't sit quite right with me. It's a bit like saying the SQL "MATCH" clause is bad because we aren't familiar with it. Our stack is Postgres presumably, rather than just the parts we're grown used to. There are many things you have to learn, I'm not sure why the boundary would be here. I'm aware this is a pattern matching reaction from me before, but I've seen a willingness to spends months and years on solutions as long as it avoided a few days reading about something new as that would be "costly".

On the flip side, I think we should definitely be questioning _why_ we need to query over the various things. This definitely shines the most when the data is a document that you want all or none of. e.g. You find by id, then get the blob, read, modify, put back. If we find we are querying stuff outside of a reporting context a lot, that might be a signal to consider going for a full DB model. This is probably for the performance and solidity the DB will bring. It's not a guarantee, but could be a signal? I don't know.

In a similar way, there's a bit of a reverse sunk cost thing going on here I think. We had some painful expectation meets reality stuff with the columns, but some of that has been paid down now.

**Comment by seanh** [comment]
Unfortunately pushing the `_safe_foo` property to Postgres (e.g. adding a Postgres-level `server_default` that sets it to an empty dict) wouldn't actually make it safe: Postgres defaults don't get applied until the DB is flushed, a new model object that hasn't been flushed yet could still have `None`. And our existing code _would_ actually crash on `GroupInfo.info` for this reason

**Comment by seanh** [comment]
I _really_ think we should back off from doing any more of this. Writing JSONSchema schemas and then our own Python code to apply those to JSON objects for writing them to Postgres seems like madness to me. Just don't use JSON in Postgres and let the _Postgres_ schema apply to the data in the usual way. This feels like we're fighting / reimplementing SQLAlchemy / Postgres to me. The way we want these JSON columns to work for us, isn't how JSON columns do work in SQLAlchemy, which makes me think that they're not intended to be used in the way that we're trying to use them.

What is it that we're actually trying to achieve by using JSON for these columns? And is there a more natural way to achieve those aims within SQLAlchemy?

**Comment by seanh** [comment]
For the `instructors` list by the way: I don't think a separate related table for it would be very difficult at all FWIW, it's just a little bit of totally standard SQLAlchemy boilerplate to add one table and relate it to another, and then it would just work with no problems. We're talking about roughly five lines of Python to add a related model class, that would just go inline in the `models/group_info.py` file. Plus the DB migration would be a line longer. It's just a [many to one relationship](https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#many-to-one). This is the most obvious and normal way to do this in SQLAlchemy so I don't think we should be afraid of it. It's a bit boilerplatey and a bit of a PITA but it's bog-standard stuff and it works.

Having said that, I think there are non-JSON column types that let you store multiple values in a single column that might not have the same problems as JSON has. E.g. the array type: https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.ARRAY

File: lms/models/group_info.py [hunk 1]
```diff
@@ -82,23 +82,31 @@ class GroupInfo(BASE):
...
 
     @property
     def _safe_info(self):
-        if self.info is None:
-            self.info = {}
+        if self._info is None:
+            self._info = {}
 
-        return self.info
+        return self._info
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/models/group_info.py:92]: The problem is that the Postgres-level default doesn't get set until the model object gets flushed to the DB. Until then, `info` is still `None`. And in the case of `GroupInfo.info` that does in fact cause our code to crash. The `ApplicationSettings` stuff has the same problem but it isn't currently causing any crashes there

**Comment by seanh** [comment]
I think my issue with JSON is that we've spent a bunch of time on it, learned a lot about how it works, and _kind of_ paid down some of the cost of dealing with its foibles, but the solution still isn't very good: the Python is still much longer and trickier than it would be with non-JSON columns, the nested mutation tracking hazard is still there. And I think we've probably bottomed out with how much time and effort it's worth spending on getting JSON-in-SQLAlchemy to behave how we think it should behave / figuring out nice and safe ways to deal with it. I don't think there's really a great way of doing it. I suspect we'll have a much easier time in future if we use multiple normal columns (for things like strings), and related tables and possibly arrays for sequences.

Mixin classes might be a good technique for things like `ApplicationSettings` too. Things like `canvas_sections_enabled` and other settings could be separate columns per individual setting on each table (`application_instance`, `course`, ...) at the Postgres level but in the Python code they could be defined once in an `ApplicationSettings` mixin class that `ApplicationInstance` and `Course` inherit from. So you can avoid repeating all the settings multiple times (in both the `ApplicationInstance` and `Course` classes) and avoid cluttering up the model classes with columns for all the settings, but still have them all be normal columns at the DB level. h does this for timestamps for example: https://github.com/hypothesis/h/blob/master/h/db/mixins.py

**Comment by seanh** [comment]
> We explicitly want to move certain aspects out of database control and into code control. Not because it makes for fewer lines of code or what-have-you, but because we want to gain freedoms from the DB.

I don't recognise the desire here. Do we want to gain freedoms from the DB? You do have to do a simple DB migration every time you add a new setting, if you're using separate columns for each setting. But apart from that I'm not sure we have a problem with being constrained by the DB here?

**Pull request merged**